### DISCUSSION TITLE
Process filename on render?

### DISCUSSION THREAD

**Discussion by allanberry** [lumeland/lume#509] [Q&A] [answered]
Hi Lume,

Thank you for beautiful software.  Really liking it so far.

So, I'm looking for a way to adjust filenames on render.  I'd like to programmatically include a version number, ideally from my site's `_data.yml` file.

To explain, I am using a somewhat strange piece of software that allows me to host individual static assets, but they're not versioned.  I'd like to know at a glance whether the assets I've uploaded are up to date.  Hence the manual versioning.

Any suggestion how this might be accomplished?

For example, something like this might work (but doesn't because it's meant for extensions):

```javascript
site.preprocess(
  ["*mah_folder_name*"],
  (page) => page.data.filename = page.src.path + "v1.2.3" + page.src.ext,
);
```

No worries if not, I'll write my own external postprocessor.  But it would be cool to manage it all in one place.

Thanks.
-Allan

**Comment by oscarotero** [comment, marked as answer]
To include the version in the url, you can use [the `url` function](https://lume.land/docs/creating-pages/urls/#urls-as-functions). For example, create `_data.js` file in the folder containing the files do you want to rewrite:

```js
// my-files/_data.js

const version = "v1.2.3";

export function url(page) {
  return `./${page.src.slug}${version}${page.src.ext}`
}
```

**Comment by allanberry** [comment]
Whoa, that's perfect.  Simpler than expected by far.  Thank you!

**Comment by allanberry** [comment]
In version 3, @oscarotero perhaps this method is no longer supported?  I can't get it to work using a `_data.js` file.  The `url()` function doesn't seem to be called at runtime.

**Comment by oscarotero** [comment]
@allanberry It's supported but there are some changes in the `page.src` object.

`page.src.slug` no longer exist. You can use the `page.data.basename` variable.
